allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,274 @@
+
+    There are several things the decoder can take from the encoder. The hidden state of the decoder can be
+    intialized with the output from the final time-step of the encoder, and the decoder may also choose to use
+    attention, in which case a weighted average of the outputs from the encoder are concatenated to the inputs
+    of the decoder at every timestep.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        Vocabulary containing ``tokens`` for input tokens and ``labels`` for output tokens.
+    source_embedder : ``TextFieldEmbedder``, required
+        Embedder for source side sequences
+    encoder : ``Seq2SeqEncoder``, required
+        The encoder of the ``EncoderDecoder`` model
+    max_decoding_steps : int, required
+        Length of decoded sequences
+    use_attention: bool
+        Should the decoder use attention to get a dynamic summary of the encoder outputs at each step of decoding?
+    scheduled_sampling_ratio: float, optional (default = 0.0)
+        At each timestep during training, we sample a random number between 0 and 1, and if it is not less than
+        this value, we use the ground truth labels. Else, we use the predictions from the previous time step.
+        If this value is 0.0 (default), this corresponds to teacher forcing, and if it is 1.0, it corresponds
+        to not using target side ground truth labels. See the following paper for more information:
+        Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks. Bengio et al., 2015.
+    """
+    def __init__(self, vocab: Vocabulary, source_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder, max_decoding_steps: int, use_attention: bool = False,
+                 scheduled_sampling_ratio: float = 0.0) -> None:
+        super(EncoderDecoder, self).__init__(vocab)
+        self._source_embedder = source_embedder
+        self._encoder = encoder
+        self._max_decoding_steps = max_decoding_steps
+        self._use_attention = use_attention
+        self._scheduled_sampling_ratio = scheduled_sampling_ratio
+        self._start_index = self.vocab.get_token_index(START_SYMBOL, "target_tokens")
+        self._end_index = self.vocab.get_token_index(END_SYMBOL, "target_tokens")
+        self._num_classes = self.vocab.get_vocab_size("target_tokens")
+        # Decoder output dim needs to be the same as the encoder output dim since we initialize the hidden
+        # state of the decoder with that of the final hidden states of the encoder. Also, if we're using
+        # attention with ``DotProductSimilarity``, this is needed.
+        self._decoder_output_dim = self._encoder.get_output_dim()
+        # TODO (pradeep): target_embedding_dim need not be the same as the source embedding dim.
+        target_embedding_dim = self._source_embedder.get_output_dim()
+        self._target_embedder = Embedding(self._num_classes, target_embedding_dim)
```

Comment:
Yes, there should be a way to let both the embeddings share weights. As a default choice, I think keeping them separate is fine because I cannot think of many cases where they should obviously be shared (summarization maybe?).
Also, I wouldn't say my toy example didn't work well. It went to a decently low NLL, and when I looked at the decoded sequences, most of them were correctly copied. Yes, it was not exactly a copy task, because it was essentially mapping symbols from the source side to a *different* set of symbols on the target side, but I am not sure if sharing the embeddings would have necessarily made it easier.

Reply:
Ok, sounds good.